### PR TITLE
[dynamo][inductor] Provide public API to get compiler options/configs

### PR DESCRIPTION
**Pull request by weifengpy**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105026

issues resolved: https://github.com/pytorch/pytorch/issues/101832

**context**: get torch.compile config for further usage. E.g, the training platform wants to get if model is compiled with cudagraph enabled and trigger further action

**how it is implemented**
   * the core logic is backend.get_compiler_config() in torch/_dynamo/eval_frame.py
   * for backend='inductor' / _TorchCompileInductorWrapper, we have inductor-specific implementation in get_compiler_config in torch/_inductor/compile_fx.py and torch/__init__.py

**how to use it**: Below is an example.

```
model = DummyModule()
optimized_module = torch.compile(
    model, options={"triton.cudagraphs": True}
)
compiler_config = optimized_module.get_compiler_config()

if compiler_config["triton.cudagraphs"]:
   pass
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

### CODE REVIEW

**Review by yanboliang — APPROVED** [review]
Looks good to me! cc @jansel for final review.

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -148,6 +149,13 @@ def __dir__(self):
             attr for attr in super().__dir__() if attr not in orig_mod_attrs
         ]
 
+    def get_compiler_config(self):
```
Comment (jansel) [inline torch/_dynamo/eval_frame.py:152]: Inductor-specific code shouldn't go in `/_dynamo/`, we should try to keep things backend agnostic.

File: torch/_dynamo/eval_frame.py [hunk 2]
```diff
@@ -148,6 +149,13 @@ def __dir__(self):
...
+        # the logic should be the same to _inductor/compile_fx.py
+        from torch._inductor import config as inductor_config
+
+        with inductor_config.patch(self.dynamo_ctx.config_patches):
+            return copy.deepcopy(inductor_config._config)
```
Comment (jansel) [inline torch/_dynamo/eval_frame.py:157]: Add a helper method to config_utils.py rather than reading private attribute.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/eval_frame.py (hunk 1 above)
Comment (willfengg) [reply torch/_dynamo/eval_frame.py]: Got you. I updated the diff with ```backend.get_compiler_config()``` so it becomes more backend agnostic

File: torch/_dynamo/eval_frame.py (hunk 2 above)
Comment (willfengg) [reply torch/_dynamo/eval_frame.py]: added helper method get_config_copy in torch/_dynamo/config_utils.py for reading private attribute

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**